xrtl/ui/control.h

```diff
@@ -0,0 +1,282 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef XRTL_UI_CONTROL_H_
+#define XRTL_UI_CONTROL_H_
+
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+
+#include "xrtl/base/macros.h"
+#include "xrtl/base/ref_ptr.h"
+#include "xrtl/base/threading/message_loop.h"
+#include "xrtl/gfx/color.h"
+
+namespace xrtl {
+namespace ui {
+
+struct Point2D {
+  int x = 0;
+  int y = 0;
+  Point2D() = default;
+  Point2D(int x, int y) : x(x), y(y) {}
+  bool operator==(const Point2D& other) const {
+    return x == other.x && y == other.y;
+  }
+  bool operator!=(const Point2D& other) const { return !(*this == other); }
+};
+
+struct Size2D {
+  int width = 0;
+  int height = 0;
+  Size2D() = default;
+  Size2D(int width, int height) : width(width), height(height) {}
+  bool operator==(const Size2D& other) const {
+    return width == other.width && height == other.height;
+  }
+  bool operator!=(const Size2D& other) const { return !(*this == other); }
+};
+
+struct Rect2D {
+  Point2D origin;
+  Size2D size;
+  Rect2D() = default;
+  Rect2D(Point2D origin, Size2D size) : origin(origin), size(size) {}
+  Rect2D(int x, int y, int width, int height)
+      : origin(x, y), size(width, height) {}
+  bool operator==(const Rect2D& other) const {
+    return origin == other.origin && size == other.size;
+  }
+  bool operator!=(const Rect2D& other) const { return !(*this == other); }
+};
+
+struct Frame2D {
+  int left = 0;
+  int top = 0;
+  int right = 0;
+  int bottom = 0;
+  Frame2D() = default;
+  Frame2D(int left, int top, int right, int bottom)
+      : left(left), top(top), right(right), bottom(bottom) {}
+  bool operator==(const Frame2D& other) const {
+    return left == other.left && top == other.top && right == other.right &&
+           bottom == other.bottom;
+  }
+  bool operator!=(const Frame2D& other) const { return !(*this == other); }
+};
+
+// A user-visible system control that can be used as a swap chain target or
+// presentation surface.
+//
+// Controls may be created and destroyed repeatedly. They will be created when
+// parented to a platform window and destroyed when unparented in addition to
```

Comment:
Controls cannot have windows - windows are the top-level objects. Controls may contain other controls, though. Not yet, so I didn't call that out here in the code, but at some point.